pkg/bgpv1/manager/reconcilerv2/preflight.go

```diff
@@ -0,0 +1,196 @@
+		shouldRecreate = true
+		l.Infof("Virtual router %s router ID has changed from %v to %v", p.DesiredConfig.Name, bgpInfo.Global.RouterID, routerID)
+	}
+	if !shouldRecreate {
+		l.Debugf("No preflight reconciliation necessary for virtual router %s", p.DesiredConfig.Name)
+		return nil
+	}
+
+	l.Infof("Recreating virtual router %s for changes to take effect", p.DesiredConfig.Name)
+	globalConfig := types.ServerParameters{
+		Global: types.BGPGlobal{
+			ASN:        uint32(localASN),
+			RouterID:   routerID,
+			ListenPort: localPort,
+			RouteSelectionOptions: &types.RouteSelectionOptions{
+				AdvertiseInactiveRoutes: true,
+			},
+		},
+	}
+
+	// stop the old BGP instance
+	p.BGPInstance.Router.Stop()
+
+	// create a new one via BGPInstance constructor
+	s, err := instance.NewBGPInstance(ctx, log, globalConfig)
+	if err != nil {
+		l.WithError(err).Errorf("Failed to start BGP server for virtual router %s", p.DesiredConfig.Name)
+		return fmt.Errorf("failed to start BGP server for virtual router %s: %w", p.DesiredConfig.Name, err)
+	}
+
+	// replace the old underlying instance with our recreated one
+	p.BGPInstance.Router = s.Router
+
+	// dump the existing config so all subsequent reconcilers perform their
+	// actions as if this is a new instance.
+	p.BGPInstance.Config = nil
+
+	// Clear the shadow state since any advertisements will be gone now that the instance has been recreated.
+	p.BGPInstance.Metadata = make(map[string]any)
+
+	return nil
+}
+
+// getLocalASN returns the local ASN for the given BGP instance. If the local ASN is defined in the desired config, it
+// will be returned. Currently, we do not support auto-ASN assignment, so if the local ASN is not defined in the
+// desired config, an error will be returned.
+func (r *PreflightReconciler) getLocalASN(p ReconcileParams) (int64, error) {
+	if p.DesiredConfig.LocalASN != nil {
+		return *p.DesiredConfig.LocalASN, nil
+	}
+
+	return -1, fmt.Errorf("missing ASN in desired config")
+}
+
+// getRouterID returns the router ID for the given ASN. If the router ID is defined in the desired config, it will
+// be returned. Otherwise, the router ID will be resolved from the ciliumnode annotations. If the router ID is not
+// defined in the annotations, the node IP from cilium node will be returned.
+func (r *PreflightReconciler) getRouterID(p ReconcileParams, asn int64) (string, error) {
+	if p.DesiredConfig.RouterID != nil {
+		return *p.DesiredConfig.RouterID, nil
+	}
+
+	// parse Node annotations into helper Annotation map
+	annoMap, err := agent.NewAnnotationMap(p.CiliumNode.Annotations)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse Node annotations for virtual router %v: %w", p.DesiredConfig.Name, err)
+	}
+
+	routerID, err := annoMap.ResolveRouterID(asn)
+	if err != nil {
+		if nodeIP := p.CiliumNode.GetIP(false); nodeIP == nil {
+			return "", fmt.Errorf("failed to get ciliumnode IP %v: %w", nodeIP, err)
+		} else {
+			routerID = nodeIP.String()
+		}
+	}
```

Comment:
Let's add a test for this auto detection case.

Reply:
Added a test case to pick router ID from node IP which is derived from CiliumNode object.